ENH: Replace Fortran COBYLA with Python version from PRIMA

nbelakovski opened the pull request:
PRIMA now has a pure python implementation of COBYLA!

I'm pleased to announce that after a lot of effort the PRIMA COBYLA algorithm is now implemented in Python. A considerable amount of effort went into validating this implementation against the Fortran and I'm happy to report that against a battery of over 500 problems from cutest, the Fortran and Python implementations get the same result bit for bit.

#### Reference issue
Closes gh-18118

#### Additional information
This commit also
- Updates COBYLA to accept the new style constraints
- Adds several options, including ftarget and more granular iprint
- Updates COBYLA to use actual bounds and linear constraints instead of simply adding them to the nonlinear constraint
- Updates COBYLA to use the new style callback interface
- Updates COBYLA to be threadsafe
- Updates all relevant documentation and tests
- Removes the old implementation

nbelakovski committed:
ENH: Replace Fortran COBLYA with Python version from PRIMA

PRIMA now has a pure python implementation of COBYLA!

This commit also
- Updates COBYLA to accept the new style constraints
- Adds several options, including ftarget and more granular iprint
- Updates COBYLA to use actual bounds and linear constraints instead of
  simply adding them to the nonlinear constraint
- Updates COBYLA to use the new style callback interface
- Updates COBYLA to be threadsafe
- Updates all relevant documentation
- Removes the old implementation

lucascolley commented:
looks like just one real CI failure from a brief look:

```
____________________________ TestBounds.test_basic _____________________________
[gw1] linux -- Python 3.10.12 /usr/bin/python3-dbg

self = <scipy.optimize.tests.test_cobyla.TestBounds object at 0x7f006a4c9f90>

    def test_basic(self):
        def f(x):
            return np.sum(x**2)
    
        lb = [-1, None, 1, None, -0.5]
        ub = [-0.5, -0.5, None, None, -0.5]
        bounds = [(a, b) for a, b in zip(lb, ub)]
        # these are converted to Bounds internally
    
        res = minimize(f, x0=[1, 2, 3, 4, 5], method='cobyla', bounds=bounds)
        ref = [-0.5, -0.5, 1, 0, -0.5]
        assert res.success
>       assert_allclose(res.x, ref, atol=1e-3)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=0.001
E       
E       Mismatched elements: 2 / 5 (40%)
E       Max absolute difference among violations: 1.85025255
E       Max relative difference among violations: 0.25694824
E        ACTUAL: array([-0.5     , -0.5     ,  1.256948,  1.850253, -0.5     ])
E        DESIRED: array([-0.5, -0.5,  1. ,  0. , -0.5])

bounds     = [(-1, -0.5), (None, -0.5), (1, None), (None, None), (-0.5, -0.5)]
f          = <function TestBounds.test_basic.<locals>.f at 0x7f0057df9440>
lb         = [-1, None, 1, None, -0.5]
ref        = [-0.5, -0.5, 1, 0, -0.5]
res        =  message: Return from COBYLA because the trust region radius reaches its lower bound.
 success: True
  status: 0
     fun: 5.753353369726385
       x: [-5.000e-01 -5.000e-01  1.257e+00  1.850e+00 -5.000e-01]
    nfev: 38
   maxcv: 0.0
self       = <scipy.optimize.tests.test_cobyla.TestBounds object at 0x7f006a4c9f90>
ub         = [-0.5, -0.5, None, None, -0.5]

```

nbelakovski commented:
There was that one, which has more to do with the fact that COBYLA seems to prioritize finding a point that satisfies the constraints over one that minimizes the objective function, but I've given it a larger starting radius and we'll see if that helps.

And there's also test_basinhopping. It takes longer now than the 20s allotted for it, taking up to 33.8 seconds on Windows fail slow full. Obviously since the algorithm is now in Python it's not surprising that it's slower, so I've increased the window for that test to 35s.

Besides those two there are some mypy failure which I'm not super familiar with so it might take me a minute to review those.

nbelakovski committed:
DEV: Don't check submodules in check_python_h_fist

nbelakovski commented:
TODO: I've added code to eliminate fixed bounds, but forgot that the code path in which I added it is not the code path through which scipy accesses the function, therefore those tests are still failing. I will move the bounds check elsewhere so that both pyprima and scipy go through the same activation path.

nbelakovski commented:
Fixed bounds and test_basinhopping are fixed, last up is the mypy failures.

lucascolley commented:
please just apply this with your next round of changes:

```diff
diff --git a/mypy.ini b/mypy.ini
index 95deab150a..3c5dfd9e86 100644
--- a/mypy.ini
+++ b/mypy.ini
@@ -751,3 +751,6 @@ ignore_errors = True
 
 [mypy-scipy._lib.array_api_compat.*]
 ignore_errors = True
+
+[mypy-scipy._lib.pyprima.*]
+ignore_errors = True

```

lucascolley commented on .gitmodules:
```diff
@@ -32,3 +32,7 @@
 	shallow = true
+[submodule "scipy/_lib/prima"]
+	path = scipy/_lib/prima
+	url = https://github.com/nbelakovski/prima
```
I suppose we will want to mirror this repo under the SciPy org?

rgommers replied:
Either that, or perhaps better have a repo with a script to extract only the relevant parts from upstream. `prima` is a very heavy repo, with a large majority of the content being benchmarks and code in other languages. So we could probably keep things a lot leaner without too much process overhead if we just copied out `pyprima/` and ensured we had a good way to record the exact commit hash and repo the content came from.

rgommers replied:
I'd be happy to look into this more if this sounds like the way to go, to avoid giving @nbelakovski even more homework. Getting PRIMA into the current great shape has already been a massive lift I think.

lucascolley replied:
Can probably use a similar script to https://github.com/lucascolley/scikit-learn/blob/xpx/maint_tools/vendor_array_api_extra.sh

dschmitz89 replied:
As a first step to get this merged, could we simply fork this repo under the scipy org?

lucascolley replied:
I can take a look at the script approach later today

lucascolley replied:
@nbelakovski https://github.com/nbelakovski/scipy/pull/1 passes tests for me locally. Would you mind taking a look and merging? @rgommers, what are the next steps to move that repo under the `scipy` org?

lucascolley commented on doc/source/tutorial/optimize.rst:
```diff
@@ -536,10 +536,10 @@ Constrained minimization
 the  constraints to be defined as a sequence of objects :func:`LinearConstraint` and
-:func:`NonlinearConstraint`. Methods ``'SLSQP'`` and ``'COBYLA'``, on the other hand,
-require constraints to be defined as a sequence of dictionaries, with keys
+:func:`NonlinearConstraint`. Methods ``'SLSQP'``, on the other hand,
+requires constraints to be defined as a sequence of dictionaries, with keys
```
nit

```suggestion
:func:`NonlinearConstraint`. Method ``'SLSQP'``, on the other hand,
requires constraints to be defined as a sequence of dictionaries, with keys
```

lucascolley commented on scipy/optimize/tests/test__basinhopping.py:
```diff
@@ -210,7 +210,7 @@ def test_all_minimizers(self):
             assert_almost_equal(res.x, self.sol[i], self.tol)
 
-    @pytest.mark.fail_slow(20)
+    @pytest.mark.fail_slow(35)
```
do we have benchmarks for COBYLA to investigate the performance impacts a bit deeper?

nbelakovski replied:
I ran some benchmarks and commented on the results here: https://github.com/scipy/scipy/pull/22350#issuecomment-2601183067

I haven't benchmarked time.

lucascolley commented on scipy/optimize/tests/test_cobyla.py:
```diff
@@ -1,17 +1,19 @@
+from numpy.testing import assert_allclose, assert_, assert_array_almost_equal
 
-from scipy.optimize import fmin_cobyla, minimize, Bounds
+from scipy.optimize import (fmin_cobyla, minimize, Bounds, NonlinearConstraint,
+    LinearConstraint, OptimizeResult)
```
optional nit in case you are pushing other things already

```suggestion
from scipy.optimize import (
    fmin_cobyla, minimize, Bounds, NonlinearConstraint, LinearConstraint,
    OptimizeResult,
)
```

lucascolley commented on scipy/optimize/tests/test_cobyla.py:
```diff
@@ -40,55 +41,81 @@ def __call__(self, x):
+            err_msg="Last design vector sent to the callback is not equal to"
+                 " returned value.",
         )
+        assert_(sol_new.success, sol_new.message)
+        assert_(sol.fun == sol_new.fun)
+        assert_(sol.maxcv == sol_new.maxcv)
+        assert_(sol.nfev == sol_new.nfev)
+        assert_(callback.n_calls == callback_new_syntax.n_calls,
+            "Callback is not called the same amount of times for old and new syntax.")
```
can you just use plain pytest `assert` here and elsewhere rather than `assert_`? See https://github.com/scipy/scipy/issues/8583

nbelakovski replied:
Done

lucascolley reviewed (commented):
just a cursory review - I haven't audited the changes to the tests in detail - but looks very promising, thanks!

nbelakovski commented:
I ran the DFO (Derivative Free Optimization) benchmarks to compare against the existing COBYLA implementation and here are the results

| Change   |   Before [2299d444-main] |   After [7ebb473d-pyprima] |         Ratio | Benchmark (Parameter)                                  |
|----------|---------------------------------|------------------------------|---------------|--------------------------------------------------------|
| +        |                     1.57221e+12 |                  6.85374e+12 |   4.36        | optimize.BenchDFO.track_all(46, 'COBYLA', 'min_obj')   |
| +        |                 85822.2         |                  2.91279e+11 |   3.39398e+06 | optimize.BenchDFO.track_all(27, 'COBYLA', 'min_obj')   |
| +        |                    31           |               1000           |  32.26        | optimize.BenchDFO.track_all(25, 'COBYLA', 'mean_nfev') |
| +        |                     4.47608e+12 |                  1.1284e+13  |   2.52        | optimize.BenchDFO.track_all(50, 'COBYLA', 'min_obj')   |
| +        |                     7.18191e+06 |                  1.05009e+09 | 146.21        | optimize.BenchDFO.track_all(17, 'COBYLA', 'min_obj')   |
| +        |                   201           |                339           |   1.69        | optimize.BenchDFO.track_all(47, 'COBYLA', 'mean_nfev') |
| +        |                   109           |                159           |   1.46        | optimize.BenchDFO.track_all(3, 'COBYLA', 'mean_nfev')  |
| +        |                   268           |                381           |   1.42        | optimize.BenchDFO.track_all(48, 'COBYLA', 'mean_nfev') |
| +        |                   149           |                207           |   1.39        | optimize.BenchDFO.track_all(45, 'COBYLA', 'mean_nfev') |
| +        |                   698           |                960           |   1.38        | optimize.BenchDFO.track_all(39, 'COBYLA', 'mean_nfev') |
| +        |                     2.21563     |                  2.86989     |   1.3         | optimize.BenchDFO.track_all(35, 'COBYLA', 'min_obj')   |
| +        |                   573           |                732           |   1.28        | optimize.BenchDFO.track_all(38, 'COBYLA', 'mean_nfev') |
| +        |                   369           |                469           |   1.27        | optimize.BenchDFO.track_all(49, 'COBYLA', 'mean_nfev') |
| +        |                   826           |               1000           |   1.21        | optimize.BenchDFO.track_all(41, 'COBYLA', 'mean_nfev') |
| +        |                   848           |               1000           |   1.18        | optimize.BenchDFO.track_all(40, 'COBYLA', 'mean_nfev') |
| +        |                   551           |                622           |   1.13        | optimize.BenchDFO.track_all(26, 'COBYLA', 'mean_nfev') |
| +        |                   882           |               1000           |   1.13        | optimize.BenchDFO.track_all(27, 'COBYLA', 'mean_nfev') |
| +        |                   413           |                459           |   1.11        | optimize.BenchDFO.track_all(29, 'COBYLA', 'mean_nfev') |
| +        |                     0.0825842   |                  0.0916973   |   1.11        | optimize.BenchDFO.track_all(44, 'COBYLA', 'min_obj')   |
| +        |                     0.290312    |                  0.316521    |   1.09        | optimize.BenchDFO.track_all(36, 'COBYLA', 'min_obj')   |
| +        |                    18.1544      |                 19.7105      |   1.09        | optimize.BenchDFO.track_all(7, 'COBYLA', 'min_obj')    |
| +        |                     1.02022e+06 |                  1.07136e+06 |   1.05        | optimize.BenchDFO.track_all(52, 'COBYLA', 'min_obj')   |
| -        |                     0.0772566   |                  0.0730388   |   0.95        | optimize.BenchDFO.track_all(43, 'COBYLA', 'min_obj')   |
| -        |                    52.6424      |                 49.7236      |   0.94        | optimize.BenchDFO.track_all(13, 'COBYLA', 'min_obj')   |
| -        |                     0.0196931   |                  0.0185623   |   0.94        | optimize.BenchDFO.track_all(20, 'COBYLA', 'min_obj')   |
| -        |                   256.166       |                241.953       |   0.94        | optimize.BenchDFO.track_all(25, 'COBYLA', 'min_obj')   |
| -        |                     0.00041551  |                  0.000381363 |   0.92        | optimize.BenchDFO.track_all(16, 'COBYLA', 'min_obj')   |
| -        |                     0.0106998   |                  0.00983711  |   0.92        | optimize.BenchDFO.track_all(18, 'COBYLA', 'min_obj')   |
| -        |                   208           |                186           |   0.89        | optimize.BenchDFO.track_all(1, 'COBYLA', 'mean_nfev')  |
| -        |                  1000           |                868           |   0.87        | optimize.BenchDFO.track_all(35, 'COBYLA', 'mean_nfev') |
| -        |                   384           |                333           |   0.87        | optimize.BenchDFO.track_all(37, 'COBYLA', 'mean_nfev') |
| -        |                   127           |                110           |   0.87        | optimize.BenchDFO.track_all(5, 'COBYLA', 'mean_nfev')  |
| -        |                     1.83741e-06 |                  1.56566e-06 |   0.85        | optimize.BenchDFO.track_all(28, 'COBYLA', 'min_obj')   |
| -        |                     0.00169863  |                  0.00140562  |   0.83        | optimize.BenchDFO.track_all(34, 'COBYLA', 'min_obj')   |
| -        |                   330           |                272           |   0.82        | optimize.BenchDFO.track_all(28, 'COBYLA', 'mean_nfev') |
| -        |                     0.0807426   |                  0.0658221   |   0.82        | optimize.BenchDFO.track_all(42, 'COBYLA', 'min_obj')   |
| -        |                    67.931       |                 54.5526      |   0.8         | optimize.BenchDFO.track_all(12, 'COBYLA', 'min_obj')   |
| -        |                  1000           |                797           |   0.8         | optimize.BenchDFO.track_all(18, 'COBYLA', 'mean_nfev') |
| -        |                  1000           |                793           |   0.79        | optimize.BenchDFO.track_all(14, 'COBYLA', 'mean_nfev') |
| -        |                   581           |                457           |   0.79        | optimize.BenchDFO.track_all(31, 'COBYLA', 'mean_nfev') |
| -        |                   174           |                132           |   0.76        | optimize.BenchDFO.track_all(0, 'COBYLA', 'mean_nfev')  |
| -        |                    91           |                 66           |   0.73        | optimize.BenchDFO.track_all(4, 'COBYLA', 'mean_nfev')  |
| -        |                     0.0230739   |                  0.0163344   |   0.71        | optimize.BenchDFO.track_all(48, 'COBYLA', 'min_obj')   |
| -        |                  1000           |                699           |   0.7         | optimize.BenchDFO.track_all(16, 'COBYLA', 'mean_nfev') |
| -        |                  1000           |                693           |   0.69        | optimize.BenchDFO.track_all(19, 'COBYLA', 'mean_nfev') |
| -        |                     0.0453879   |                  0.0311729   |   0.69        | optimize.BenchDFO.track_all(51, 'COBYLA', 'min_obj')   |
| -        |                   105           |                 69           |   0.66        | optimize.BenchDFO.track_all(2, 'COBYLA', 'mean_nfev')  |
| -        |                     0.0342853   |                  0.0215811   |   0.63        | optimize.BenchDFO.track_all(49, 'COBYLA', 'min_obj')   |
| -        |                     0.155084    |                  0.0939493   |   0.61        | optimize.BenchDFO.track_all(23, 'COBYLA', 'min_obj')   |
| -        |                     0.0138684   |                  0.00717343  |   0.52        | optimize.BenchDFO.track_all(19, 'COBYLA', 'min_obj')   |
| -        |                  1000           |                462           |   0.46        | optimize.BenchDFO.track_all(8, 'COBYLA', 'mean_nfev')  |
| -        |                     0.102238    |                  0.046054    |   0.45        | optimize.BenchDFO.track_all(22, 'COBYLA', 'min_obj')   |
| -        |                    48           |                 21           |   0.44        | optimize.BenchDFO.track_all(17, 'COBYLA', 'mean_nfev') |
| -        |                     9.46318e-06 |                  3.81706e-06 |   0.4         | optimize.BenchDFO.track_all(29, 'COBYLA', 'min_obj')   |
| -        |                     0.0202994   |                  0.00759472  |   0.37        | optimize.BenchDFO.track_all(45, 'COBYLA', 'min_obj')   |
| -        |                     0.0635923   |                  0.0203449   |   0.32        | optimize.BenchDFO.track_all(6, 'COBYLA', 'min_obj')    |
| -        |                     0.193012    |                  0.0575186   |   0.3         | optimize.BenchDFO.track_all(21, 'COBYLA', 'min_obj')   |
| -        |                     0.00311681  |                  0.000732976 |   0.24        | optimize.BenchDFO.track_all(10, 'COBYLA', 'min_obj')   |
| -        |                     0.0252814   |                  0.00546931  |   0.22        | optimize.BenchDFO.track_all(47, 'COBYLA', 'min_obj')   |
| -        |                     1.43505     |                  0.281975    |   0.2         | optimize.BenchDFO.track_all(9, 'COBYLA', 'min_obj')    |
| -        |                  1000           |                186           |   0.19        | optimize.BenchDFO.track_all(24, 'COBYLA', 'mean_nfev') |
| -        |                     0.00629085  |                  0.00105276  |   0.17        | optimize.BenchDFO.track_all(11, 'COBYLA', 'min_obj')   |
| -        |                     0.00595643  |                  0.000906142 |   0.15        | optimize.BenchDFO.track_all(8, 'COBYLA', 'min_obj')    |
| -        |                     0.0024656   |                  3.83688e-05 |   0.02        | optimize.BenchDFO.track_all(24, 'COBYLA', 'min_obj')   |
| -        |                     8.37489     |                  0.0280712   |   0           | optimize.BenchDFO.track_all(15, 'COBYLA', 'min_obj')   |


43 problems are improved either in final value or in number of function evaluations. 22 are worse in those metrics. A notable outlier is test #27. I did some digging and found this this is the objective function it is trying to evaluate

```
def func(x):
    m = 20
    fvec = np.zeros(m)
    for i in range(m):
            temp = (i + 1) / 5
            tmp1 = x[0] + temp * x[1] - np.exp(temp)
            tmp2 = x[2] + np.sin(temp) * x[3] - np.cos(temp)
            fvec[i] = tmp1 * tmp1 + tmp2 * tmp2
    y = np.sum(fvec ** 2)
    if np.isnan(y):
        return np.inf
    return y
x0 = [250.  50. -50. -10.]
```

If I remove the maxfun limitation on the new implementation it eventually gets to the optimum value, although it takes ~13000 function evaluations on my machine as compared to <1000 for the old implementation. Problem 17 is also somewhat of an outlier. If I tighten the tolerance on it to 1e-6 from the default 1e-4 the result becomes comparable to the old implementation. Different problems will fare better or worse depending on the initial/final trust region radius (which in scipy are called rhobeg and tol respectively) and these parameters are exposed to the user so I don't think this outlier is something to be overly concerned with.

nbelakovski committed:
TST: Remove usage of assert_ from test_cobyla

dschmitz89 commented:
For me the benchmarks look like a net win. Let's include one sentence about the parameters that can be tweaked in the release notes to give users a hint.

nbelakovski committed:
DOC: Update COBYLA docs

It no longer wraps a Fortran implementation. I'm not sure why the line
about the function returning a list of number was there in the first
place. It doesn't make much sense, and also doesn't work as of this
writing with the existing implementation and certainly not with the new
one.

nbelakovski committed:
DOC: Add release notes for COBYLA

ilayn commented:
Not to block, or in fact take away, anything from this great addition, I just want to understand, is this going to live in your fork and not in prima repo? I don't have any background info hence my question.

nbelakovski commented:
> Not to block, or in fact take away, anything from this great addition, I just want to understand, is this going to live in your fork and not in prima repo? I don't have any background info hence my question.

At the moment yes. It's my understanding that @zaikunzhang is occupied and that PRIMA is not particularly high on his list of priorities. As such I haven't had a chance to discuss this with him and see if he has any comments, but given my previous work with him, I think he would take it as a good sign that the Python implementation is bit-for-bit identical* to the Fortran version on all problems in the test set that we used for the Python bindings. Hopefully he is able to get bandwidth sometime in the future and we'll be able to merge the Python implementation back into the prima repo. I've been rebasing my work on the latest stuff from him so that the moment the fork is up to date with the latest. All that said, I don't think we should wait for him to move forward on this.

\* This is true when the USE_NAIVE_MATH flag in Python is on, and Fortran is compiled in debug mode. I've chosen to default that flag to off so that numpy routines are used for martix multiplication and so forth instead of the naive implementations. My reasoning is that most practical problems do not need to preserve all floating point calculations, but when the flag is on it sometimes results in more function evaluations and therefore more matrix math, and so the speedup/slowdown is not as significant as you might think. I'm open to changing this.

nbelakovski committed:
TST: Give a little more time for the basinhopping test [wheel build]

nbelakovski commented:
Looks like it's failing on `test_mpmath.py::TestSystematic::test_besselj_complex`, but it looks like that's failing in [this recently closed PR](https://github.com/scipy/scipy/actions/runs/13030267545/job/36347803186?pr=22240) as well, so unrelated to these code changes (also the test is not related to optimization of any kind).

rgommers commented:
It looks like this is pretty happy. It'd be useful to push a commit with `[wheel build]` on the first line of the commit message @nbelakovski, if you didn't already do that recently, in order to verify that all wheel builds are happy too (test suites on, e.g., musllinux can sometimes turn up new failures).

nbelakovski commented:
The wheel build seems rather flaky. The first time I ran it I got two errors, both of which were due to libmamba failing to extract something. I tried rebasing on various commits that had successful wheel builds, but then I got other errors. Finally I rebased on the latest version of main as of this writing, and I have

- A docker failure on cp313-manylinux aarch64
- A libmamba failure when trying to extract openmp_mutex on cp312-musllinux aarch64 
- 5 failures on cp3[10, 11, 12, 13, 13t]-macosx x86_64 openblas 10.13 due to a failure to compile `scipy/sparse/csgraph/_shortest_path.cpython-310-darwin.so.p/meson-generated__shortest_path.cpp`

I'm not sure why these are failing but none of them appear related to any changes in this PR.

dschmitz89 commented:
@rgommers : is there anything left to update here in your opinion? The wheel failures on macOS 10 are tracked in #22487 , the other wheel failures are flaky CI hickups.

rgommers commented:
The churn sometimes seems to be never-ending.🤦🏼  Please ignore in this PR, CI doesn't have to be green for merging.

nbelakovski commented:
> The churn sometimes seems to be never-ending.🤦🏼 Please ignore in this PR, CI doesn't have to be green for merging.

Indeed. I just rebased to get the fix for actions/cache but I think I'll stop here in terms of trying to get the CI all green. The 3 remaining failures are spurious and not related to the code, and so I guess the only thing left to do is vendoring (by which I assuming you mean creating scipy/prima?)

nbelakovski commented:
I know you guys have lots of priorities and this one probably isn't on the top of the list, but is there any chance you might be able to get to it soon? It would be really great to officially mark it done and make some relevant announcements!

lucascolley committed:
MAINT: remove prima submodule

lucascolley committed:
MAINT: add pyprima submodule

lucascolley committed:
BLD: update meson for pyprima

lucascolley committed:
switch to alternative script

nbelakovski committed:
Merge pull request #1 from lucascolley/pyprima-vendor

Vendor separate minimal repo

dschmitz89 commented:
Test failures are unrelated, this should be merge able now.

lucascolley commented:
We will want to transfer lucascolley/pyprima to the SciPy org before merge.

dschmitz89 commented:
> We will want to transfer lucascolley/pyprima to the SciPy org before merge.

Argh, missed that, sry. Maintainer should have rights to create repos under scipy org, at least the button works for me. Is there an approval process to follow?

lucascolley committed:
delete lucascolley/pyprima submodule

lucascolley committed:
add scipy/pyprima submodule

dschmitz89 commented:
Thanks @lucascolley . If there are no more comments, I will merge this next week.

dschmitz89 committed:
Merge branch 'main' into pyprima

dschmitz89 commented:
@lucascolley @nbelakovski : the new dependency checker `tach` is complaining now. Would the suggested fix by tach suffice?

lucascolley commented:
I would suggest just adding

```toml
[[modules]]
path = "scipy._lib.pyprima"
unchecked = true
```

nbelakovski committed:
Remove tach check on PRIMA

PyPRIMA imports scipy in order to use the constraint objects, which
results in a circular dependency violation. COBYQA does the same thing,
so like with COBYQA we set unchecked to true.

nbelakovski commented:
Agreed, and done.

dschmitz89 commented:
Alright, I pulled the trigger here finally.

Thanks for your tremendous work here @nbelakovski ! Hat tip to all reviewers as well.

j-bowhay commented:
And of course thanks for your great work @nbelakovski

nbelakovski commented:
Thanks all, and thanks for the help in getting it across the finish line!

zaikunzhang commented:
(sorry for reposting)

Hello! Author of PRIMA here.

Thanks for working on this. I am sorry for not being able to respond.

For anything derived from PRIMA, may I request that the following be stated clearly?

PRIMA is developed by Zaikun Zhang, with "Zaikun Zhang" linked to my homepage [zhangzk.net](https://www.zhangzk.net/)
PRIMA means "Reference Implementation for Powell's Methods with Modernization and Amelioration", which provides a modernized and improved implementation of M.J.D. Powell's derivative-free optimization methods, with "M.J.D. Powell" linked to [zhangzk.net/powell.html](https://www.zhangzk.net/powell.html)
As an example, see [lfortran.org/blog/2025/03/lfortran-compiles-prima](https://lfortran.org/blog/2025/03/lfortran-compiles-prima/)

Note that PRIMA contains [bug fixes](https://github.com/libprima/prima?rgh-link-date=2025-03-25T02%3A57%3A19.000Z#bug-fixes) and [improvements](https://github.com/libprima/prima?rgh-link-date=2025-03-25T02%3A57%3A19.000Z#improvements) that do not exist in Powell's original implementation of the solvers. Results produced by PRIMA may not be reproducible using the original solvers. Therefore, it is simply wrong.

It took me years to develop PRIMA, which almost cost my career and my life. (This partially explains why I have been unable to respond recently. I have to strive to survive as a young professor).

As a professor, [all I need is a citation / acknowledgment when my work is useful](https://github.com/libprima/prima?rgh-link-date=2025-03-25T02%3A57%3A19.000Z#citing-prima), which is not well recognized in academia when it comes to software development. In addition, for obvious reasons, people tend to cite Powell without mentioning me when they use PRIMA instead of Powell's original solvers. (I have no doubt that my contribution is insignificant compared with Powell's, and I acknowledge this whenever I introduce PRIMA to others. See [libprima.net](https://www.libprima.net/) for example). This is crucial for my career.

Thank you very much.

rgommers commented:
@dschmitz89 and @lucascolley thank you for finishing the review/merge here! Sorry for going missing in action for a while. 

@nbelakovski awesome work, thank you 🙏🏼 

> As a professor, [all I need is a citation / acknowledgment when my work is useful](https://github.com/libprima/prima?rgh-link-date=2025-03-25T02%3A57%3A19.000Z#citing-prima), which is not well recognized in academia when it comes to software development.

We definitely recognize that problem and the need for citations. I know you saw it, but just for completeness: gh-22738 added references to your paper in two places.
